### PR TITLE
feat: vm.writeJson()

### PR DESCRIPTION
**Pull request by odyslam**
## Motivation

We want to store json files as part of chain-ops operations

## Solution

- Introduce `vm.serializeJson()` and `vm.writeJson()` cheatcodes
- When the user serializes a value, they pass a key for the object and a key for the value. The function returns a stringified form of the JSON up to that moment. That means that users can use this to create arbitrarily nested JSON files. The cheatcode will interpret the stringified JSON as a JSON object.
e.g 

```Solidity
vm.serializeUint(json1, "key1", uint(254));
vm.serializeBool(json1, "boolean", true);
vm.serializeInt(json2, "key2", -234);
vm.serializeUint(json2,"deploy", uint(254));
string memory data = vm.serializeBool(json2,"boolean", true);
vm.serializeString(json2, "json1", data);
string memory path = "./output/artifact.json";
vm.writeJson(path, json1); // or vm.writeJson(path, json1, ".nestedObject");
```

- When the user wants to write the JSON, they can use the same object key to write it:
	- either to a new file
	- or directly edit the value of an existing JSON file. This enables the user to update values without having to parse the whole configuration json (which can be unwieldy)
- The `vm.serialize*` cheatcodes will be used indirectly via stdJSON library, which will offer helpers so that user can do:
```Solidity
string memory json1 = "key";
Serializer s = new Serializer(json1);
s.serialize("value", true);
string memory object = s.serialize("number", 123);
s.serialize("object1", object);
...
```
- The said helpers will be added to `forge-std` in a later PR, alongside the appropriate docs
- Thank you @mattsse for helping be find the bearings when I started this feature 💕

### CODE REVIEW

**Comment by odyslam** [comment]
I am getting a bunch of Clippy errors from irrelevant parts of the codebase. Should I address them?

**Comment by mattsse** [comment]
> I am getting a bunch of Clippy errors from irrelevant parts of the codebase. Should I address them?

they could be fixed on master already, if not, just ignore

File: evm/src/executor/abi/mod.rs
Comment (mattsse) [inline evm/src/executor/abi/mod.rs:138]: what's the first argument here?
the json path?

**Review by mattsse — COMMENTED** [review]
really like this.

the thing that could be tricky is nested objects, passing the `object_key` is not a big deal but also not ideal.
If I understood that correctly how it works.

File: evm/src/executor/abi/mod.rs
Comment (odyslam) [reply evm/src/executor/abi/mod.rs:138]: object key, value key, value

File: evm/src/executor/inspector/cheatcodes/ext.rs
Comment (mattsse) [inline evm/src/executor/inspector/cheatcodes/ext.rs:347]: fix lint issue, `json_str` is already a reference

File: evm/src/executor/inspector/cheatcodes/ext.rs
Comment (mattsse) [inline evm/src/executor/inspector/cheatcodes/ext.rs:365]: please document the new functions

File: evm/src/executor/inspector/cheatcodes/ext.rs
Comment (mattsse) [inline evm/src/executor/inspector/cheatcodes/ext.rs:366]: rm leading `_` which usually indicates variable is not used.

File: testdata/cheats/Cheats.sol
Comment (mattsse) [inline testdata/cheats/Cheats.sol:230]: okay, I've understood how this works now.

providing an identifier key is not really ideal imo.

In order to serialize nested objects, you need to provide the path yourself, right?

This also limits serializing/deserializing to json only, right?

Having a way to abstract this away would be ideal.

that's what I tried with the `Serializer` approach, which would also make the identifier value unnecessary, but is more effort to implement.

I'm hesitant to merge this as is and then make breaking changes later.

I really would like a special json Serializer contract instead.

Internally all the serialize functions will work similar, but this way we can get rid of this identifier, if for example a `jsonSerializer() returns Serializer` cheat code returns a contract which case a new internal id, so instead of `vm.serializeAddress(id, key, value)` you can just call `serializer.serializeAddress(key, value)`


But perhaps I'm overthinking this and this is good enough, wdyt @gakonst?

**Review by mattsse — CHANGES_REQUESTED** [review]

File: testdata/cheats/Cheats.sol
Comment (odyslam) [reply testdata/cheats/Cheats.sol:230]: I agree100%. The design assumes this contract to be included in forge-std:

```
contract Serializer {

    string jsonKey;

    address constant HEVM_ADDRESS = address(bytes20(uint160(uint256(keccak256("hevm cheat code")))));

    Cheats vm = Cheats(HEVM_ADDRESS);

    constructor(string memory key){
        jsonKey = key;
    }

    function serialize(string calldata key, bool value) external returns(string memory){
        return vm.serializeBool(jsonKey, key, value);
    }

    function serialize(string calldata key, bool[] memory value) external returns(string memory){
        return vm.serializeBool(jsonKey, key, value);
    }


    function serialize(string calldata key, uint256 value)external returns(string memory){
        return vm.serializeUint(jsonKey, key, value);
    }

    function serialize(string calldata key, uint256[] memory value)external returns(string memory){
        return vm.serializeUint(jsonKey, key, value);
    }


    function serialize(string calldata key, int256 value)external returns(string memory){
        return vm.serializeInt(jsonKey, key, value);
    }

    function serialize(string calldata key, int256[] memory value)external returns(string memory){
        return vm.serializeInt(jsonKey, key, value);
    }

    function serialize(string calldata key, address value)external returns(string memory){
        return vm.serializeAddress(jsonKey, key, value);
    }

    function serialize(string calldata key, address[] memory value)external returns(string memory){
        return vm.serializeAddress(jsonKey, key, value);
    }

    function serialize(string calldata key, bytes32 value)external returns(string memory){
        return vm.serializeBytes32(jsonKey, key, value);
    }

    function serialize(string calldata key, bytes32[] memory value)external returns(string memory){
        return vm.serializeBytes32(jsonKey, key, value);
    }

    function serialize(string calldata key, string memory value)external returns(string memory){
        return vm.serializeString(jsonKey, key, value);
    }

    function serialize(string calldata key, string[] memory value)external returns(string memory){
        return vm.serializeString(jsonKey, key, value);
    }

    function write(string calldata path) external {
        vm.writeJson(jsonKey, path);
    }
}
```

**Review by mattsse — APPROVED** [review]
I think we can merge this and iterate on additional user feedback

needs `forge fmt testdata`

**Comment by odyslam** [comment]
@mattsse Apart from the helper contract `Serializer` that will be merged into forge-std; do you have any other suggestion from a DevEx PoV?

**Comment by odyslam** [comment]
@mds1 since you are here.. I am thinking now that Serializer should actually be part of the stdJson library and follow the same pattern

instead of 
```
Serializer s = new Serializer(json1)
s.serialize(boolean, "true")
```

we should have
```
using StdJson for string;
json1 = "deploy";
json1.serialize(boolean, "true");
```

Agreed? If yes, will proceed to make a PR in `forge-std`

**Comment by mds1** [comment]
@odyslam tbh I have not read through this PR / the UX yet 😅 let me catch up on this later today and I'll share any thoughts

**Comment by mattsse** [comment]
unfortunately, this broke a couple of tests because storing json in evm memory causes evm memory allocation overflow

https://github.com/foundry-rs/foundry/actions/runs/3411552220/jobs/5676035725

**Comment by odyslam** [comment]
@mattsse 

What do you mean? How is this different from parseJson?

**Comment by mattsse** [comment]
the test started failing, looks like it was introduced by this PR

**Comment by mds1** [comment]
if I check out the commit this was merged (64cbdd183e0aae99eb1be507196b6b5d640b3801) the forge-std tests fail on master, if I check out the commit right before this was merged (01b16238ff87dc7ca8ee3f5f13e389888c2a2ee4) they pass

**Comment by odyslam** [comment]
@mds1  I changed the jsonpath library with one that supported a replace action (the previous only supported search). 

The tests in testdata passed, will see why and submit a PR to fix this

**Pull request merged**
